src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp

```diff
@@ -64,266 +68,366 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-		// In particular this has an effect when the setting is modified by 180 degrees, in which case the drone
-		// would pass above the target without the filter. The filtering makes it so that the drone flies around
-		// the target into the new postion.
-		_follow_angle_filter.setParameters(_deltatime, FOLLOW_ANGLE_FILTER_ALPHA);
-		_follow_angle_filter.update(_follow_angle_rad);
-
-		// Wrap around 360 degrees
-		if (_follow_angle_filter.getState() > M_TWOPI_F) {
-			_follow_angle_filter.reset(_follow_angle_filter.getState() - M_TWOPI_F);
-			_follow_angle_rad = _follow_angle_rad - M_TWOPI_F;
-
-		} else if (_follow_angle_filter.getState() < -M_TWOPI_F) {
-			_follow_angle_filter.reset(_follow_angle_filter.getState() + M_TWOPI_F);
-			_follow_angle_rad = _follow_angle_rad + M_TWOPI_F;
-		}
+	// Reset last valid estimator data received timestamp
+	_last_valid_target_estimator_timestamp = follow_target_estimator.timestamp;
 
-		// Assume the target's velocity vector is its heading and use it to construct the offset vector
-		// such that drone_pos_setpoint = target_pose + offset_vector
-		if (vel_ned_est.longerThan(MINIMUM_SPEED_FOR_HEADING_CHANGE) &&
-		    vel_ned_est.longerThan(FLT_EPSILON)) {
-			// Compute offset vector relative to target position. At the same time the offset vector defines the
-			// vieweing angle of the drone
-			_target_velocity_unit_vector = Vector2f(vel_ned_est.xy()).unit_or_zero();
-		}
+	// Reset the filter if any pose filter states are not finite or estimator timed out
+	if ((!PX4_ISFINITE(_target_position_velocity_filter.getState()(0)) || !PX4_ISFINITE(_target_position_velocity_filter.getState()(1))
+		|| !PX4_ISFINITE(_target_position_velocity_filter.getState()(2)) || !PX4_ISFINITE(_target_position_velocity_filter.getRate()(0))
+		|| !PX4_ISFINITE(_target_position_velocity_filter.getRate()(1)) || !PX4_ISFINITE(_target_position_velocity_filter.getRate()(2))) || target_estimator_timeout) {
+		_target_position_velocity_filter.reset(pos_ned_est, vel_ned_est);
+	}
+	else {
+		_target_position_velocity_filter.update(_deltatime, pos_ned_est, vel_ned_est);
+	}
+}
 
-		float offset_x = (float)cos(_follow_angle_filter.getState()) * _target_velocity_unit_vector(0) - (float)sin(
-					 _follow_angle_filter.getState()) * _target_velocity_unit_vector(1);
-		float offset_y = (float)sin(_follow_angle_filter.getState()) * _target_velocity_unit_vector(0) + (float)cos(
-					 _follow_angle_filter.getState()) * _target_velocity_unit_vector(1);
+void FlightTaskAutoFollowTarget::update_rc_adjusted_follow_height(const Sticks &sticks) {
+	if (fabsf(sticks.getThrottleZeroCentered()) < USER_ADJUSTMENT_DEADZONE) {
+		return;
+	}
+	// Only apply Follow height adjustment if height setpoint and current height are within time window
+	if(fabsf(_position_setpoint(2) - _position(2)) < FOLLOW_HEIGHT_USER_ADJUST_SPEED * USER_ADJUSTMENT_ERROR_TIME_WINDOW) {
```

Comment:
I don't quite understand why `USER_ADJUSTMENT_ERROR_TIME_WINDOW` is called a time window.

- On the left hand side we have `fabsf(_position_setpoint(2) - _position(2))`, which is the tracking error.
- On the right hand side we have two constant variables acting as a threshold.

Does this prevent adjustments of heigh / angle / distance, when the error is greater than what's deemed an acceptable error? Why is this necessary? I mean, what real situation is it preventing from happening?

Reply:
It's well described in the follow angle adjustment [function](https://github.com/PX4/PX4-Autopilot/pull/19260/files#diff-c427e5824011523926ec83abd52b8dfe31261a95bf66c1b12394fb79573d84fdR169) below.

If the follow characteristics are changed without taking real drone's position into consideration, the drone can end up following that 'imaginary' setpoint and user's RC input won't actually affect drone's behavior since the setpoint is so far fetched (especially the case for angle adjustments)

Reply:
Interesting, now I understand why you're doing it! 
But where does "time" come into play? I'm just hung up on the terminology.

Reply:
I found it hard to generalize the calculation of an arbitrary constraint for follow distance, angle and height.

Which is why I went with a time constant, since then it can generally provide enough buffer zone in user's perspective (max 0.5 second delay from control input to action of the drone, roughly. Does that make sense?